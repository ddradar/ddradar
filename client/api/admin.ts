import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
