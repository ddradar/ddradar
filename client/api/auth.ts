import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

type AuthResult = {
  clientPrincipal: Api.ClientPrincipal | null
}

/**
 * Call "Get Credentials" API.
 * @see https://docs.microsoft.com/azure/static-web-apps/user-information
 */
export async function getClientPrincipal(
  $http: Pick<NuxtHTTPInstance, '$get'>
) {
  const { clientPrincipal } = await $http.$get<AuthResult>('/.auth/me')
  return clientPrincipal
}
