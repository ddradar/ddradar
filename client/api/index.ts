export const apiPrefix = '/api/v1'
