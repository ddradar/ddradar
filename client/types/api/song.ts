/**
 * Object type returned by `/api/v1/songs/{:songId}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/getSongInfo/README.md
 */
export type SongInfo = {
  /**
   * Song id that depend on official site.
   * @example `^([01689bdiloqDIOPQ]*){32}$`
   */
  id: string
  name: string
  /**
   * Song furigana for sorting.
   * @example `^([A-Z0-9 .ぁ-んー]*)$`
   */
  nameKana: string
  /**
   * Index for sorting. Associated with the "Choose by Name" folder.
   * @example `0`: あ行, `1`: か行, ..., `10`: A, `11`: B, ..., `35`: Z, `36`: 数字・記号
   */
  nameIndex: number
  artist: string
  /** Series title depend on official site. */
  series: string
  /**
   * Displayed min BPM (Beet Per Minutes).
   * Set to `null` if not revealed, such as "???".
   */
  minBPM: number | null
  /**
   * Displayed max BPM (Beet Per Minutes).
   * Set to `null` if not revealed, such as "???".
   */
  maxBPM: number | null
  charts: StepChart[]
}

export type StepChart = {
  /** `1`: SINGLE, `2`: DOUBLE */
  playStyle: 1 | 2
  /**
   * `0`: BEGINNER,
   * `1`: BASIC,
   * `2`: DIFFICULT,
   * `3`: EXPERT,
   * `4`: CHALLENGE
   */
  difficulty: 0 | 1 | 2 | 3 | 4
  level: number
  /** Normal arrow count. (Jump = 1 count) */
  notes: number
  /** Freeze Arrow count. */
  freezeArrow: number
  /** Shock Arrow count. */
  shockArrow: number
  /** Groove Radar STREAM */
  stream: number
  /** Groove Radar VOLTAGE */
  voltage: number
  /** Groove Radar AIR */
  air: number
  /** Groove Radar FREEZE */
  freeze: number
  /** Groove Radar CHAOS */
  chaos: number
}

/**
 * Object type returned by `/api/v1/songs/name/{:name}` and `/api/v1/songs/series/{:series}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/searchSongByName/README.md
 * @see https://github.com/ddradar/ddradar/blob/master/api/searchSongBySeries/README.md
 */
export type SongListData = Omit<SongInfo, 'charts'>

/**
 * Object type returned by `/api/v1/charts/{:playStyle}/{:level}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/searchCharts/README.md
 */
export type ChartInfo = Pick<SongInfo, 'id' | 'name' | 'series'> &
  Pick<StepChart, 'playStyle' | 'difficulty' | 'level'>

export const SeriesList: string[] = [
  'DDR 1st',
  'DDR 2ndMIX',
  'DDR 3rdMIX',
  'DDR 4thMIX',
  'DDR 5thMIX',
  'DDRMAX',
  'DDRMAX2',
  'DDR EXTREME',
  'DDR SuperNOVA',
  'DDR SuperNOVA2',
  'DDR X',
  'DDR X2',
  'DDR X3 VS 2ndMIX',
  'DanceDanceRevolution (2013)',
  'DanceDanceRevolution (2014)',
  'DanceDanceRevolution A',
  'DanceDanceRevolution A20',
  'DanceDanceRevolution A20 PLUS',
]

export const NameIndexList: string[] = [
  'あ',
  'か',
  'さ',
  'た',
  'な',
  'は',
  'ま',
  'や',
  'ら',
  'わ',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '数字・記号',
]

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export function getPlayStyleName(
  playStyle: number
): 'SINGLE' | 'DOUBLE' | 'UNKNOWN' {
  return (['SINGLE', 'DOUBLE'] as const)[playStyle - 1] ?? 'UNKNOWN'
}

export function getDifficultyName(
  difficulty: number
): 'BEGINNER' | 'BASIC' | 'DIFFICULT' | 'EXPERT' | 'CHALLENGE' | 'UNKNOWN' {
  return (
    (['BEGINNER', 'BASIC', 'DIFFICULT', 'EXPERT', 'CHALLENGE'] as const)[
      difficulty
    ] ?? 'UNKNOWN'
  )
}
