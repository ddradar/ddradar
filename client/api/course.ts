import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function getCourseType(type: number) {
  return type === 1 ? 'NONSTOP' : type === 2 ? '段位認定' : ''
}

/**
 * Call "Get Course List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses/
 */
export function getCourseList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series?: 16 | 17,
  type?: 1 | 2
) {
  const searchParams = new URLSearchParams()
  if (series) searchParams.append('series', `${series}`)
  if (type) searchParams.append('type', `${type}`)
  return $http.$get<Api.CourseListData[]>(`${apiPrefix}/courses`, {
    searchParams,
  })
}

/**
 * Call "Get Course Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses__id/
 */
export function getCourseInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.CourseInfo>(`${apiPrefix}/courses/${id}`)
}
