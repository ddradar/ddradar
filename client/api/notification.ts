import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Notification List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification/
 */
export function getNotificationList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  topOnly?: boolean
) {
  const query = topOnly ? '?scope=top' : ''
  return $http.$get<Api.NotificationListData[]>(
    `${apiPrefix}/notification${query}`
  )
}

/**
 * Call "Get Notification Info" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification__id/
 */
export function getNotificationInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.NotificationInfo>(`${apiPrefix}/notification/${id}`)
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification--post/
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Omit<Api.NotificationBody, 'sender'>
) {
  return $http.$post<Api.NotificationBody>(`${apiPrefix}/notification`, {
    ...body,
    sender: 'SYSTEM',
  })
}
