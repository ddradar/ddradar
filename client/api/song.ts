import type { Api } from '@ddradar/core'
import { Song } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export function getChartTitle({
  playStyle,
  difficulty,
  level,
}: Pick<Api.ChartInfo, 'playStyle' | 'difficulty' | 'level'>) {
  const shortPlayStyle = `${Song.playStyleMap.get(playStyle)![0]}P` as
    | 'SP'
    | 'DP'
  const difficultyName = Song.difficultyMap.get(difficulty)!
  return `${shortPlayStyle}-${difficultyName} (${level})`
}

export function getDisplayedBPM({
  minBPM,
  maxBPM,
}: Pick<Api.SongInfo, 'minBPM' | 'maxBPM'>) {
  if (!minBPM || !maxBPM) return '???'
  if (minBPM === maxBPM) return `${minBPM}`
  return `${minBPM}-${maxBPM}`
}

/**
 * Call "Get Song Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__id/
 */
export function getSongInfo($http: Pick<NuxtHTTPInstance, '$get'>, id: string) {
  return $http.$get<Api.SongInfo>(`${apiPrefix}/songs/${id}`)
}

/**
 * Call "Search Song by Name" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__name__id/
 */
export function searchSongByName(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  nameIndex: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/name/${nameIndex}`)
}

/**
 * Call "Search Song by Series" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__series__id/
 */
export function searchSongBySeries(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/series/${series}`)
}

/**
 * Call "Search Charts" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/charts__style__level/
 */
export function searchCharts(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  playStyle: Song.PlayStyle,
  level: number
) {
  return $http.$get<Api.ChartInfo[]>(
    `${apiPrefix}/charts/${playStyle}/${level}`
  )
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs--post/
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Api.SongInfo
) {
  return $http.$post<Api.SongInfo>(`${apiPrefix}/songs`, body)
}
