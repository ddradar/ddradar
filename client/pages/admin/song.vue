<template>
  <section class="section">
    <h1 class="title">Add/Update Song</h1>
    <b-field label="Song ID">
      <b-field grouped>
        <b-input
          v-model="id"
          maxlength="32"
          required
          pattern="^[01689bdiloqDIOPQ]{32}$"
        />
        <b-button
          type="is-primary"
          :disabled="!isValidSongId"
          @click="loadSongInfo()"
        >
          Load
        </b-button>
      </b-field>
    </b-field>

    <b-field label="Name">
      <b-input v-model="name" required />
    </b-field>

    <b-field label="Furigana">
      <b-input
        v-model="nameKana"
        required
        pattern="^([A-Z0-9 .ぁ-んー]*)$"
        placeholder="A-Z 0-9 . あ-ん ー"
      />
    </b-field>

    <b-field label="Artist">
      <b-input v-model="artist" />
    </b-field>

    <b-field label="BPM">
      <b-field grouped>
        <b-input v-model.number="minBPM" type="number" placeholder="min" />
        <span class="control">-</span>
        <b-input v-model.number="maxBPM" type="number" placeholder="max" />
      </b-field>
    </b-field>

    <b-field v-for="(chart, i) in charts" :key="i" :label="`Chart ${i + 1}`">
      <b-field grouped group-multiline>
        <b-select placeholder="PlayStyle"></b-select>
        <b-select placeholder="Difficulty"></b-select>
        <b-input placeholder="Level" type="number" min="1" max="20" />
        <b-field>
          <b-input placeholder="Notes" type="number" min="0" max="9999" />
          <b-input placeholder="FA" type="number" min="0" max="9999" />
          <b-input placeholder="SA" type="number" min="0" max="9999" />
        </b-field>
        <b-field>
          <b-input placeholder="Str" type="number" min="0" max="300" />
          <b-input placeholder="Vol" type="number" min="0" max="300" />
          <b-input placeholder="Air" type="number" min="0" max="300" />
          <b-input placeholder="Fre" type="number" min="0" max="300" />
          <b-input placeholder="Cha" type="number" min="0" max="300" />
        </b-field>
      </b-field>
    </b-field>
    <b-field>
      <b-button icon-left="plus">Add Chart</b-button>
    </b-field>

    <b-field>
      <b-button type="is-success">Save</b-button>
    </b-field>
  </section>
</template>

<script lang="ts">
import { Component, Vue } from 'nuxt-property-decorator'

import { SongInfo, StepChart } from '~/types/api/song'

@Component
export default class SongEditorPage extends Vue implements SongInfo {
  id: string = ''
  name: string = ''
  nameKana: string = ''
  nameIndex: number = 0
  artist: string = ''
  series: string = ''
  minBPM: number | null = null
  maxBPM: number | null = null
  charts: StepChart[] = [
    {
      playStyle: 1,
      difficulty: 0,
      level: 3,
      notes: 67,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 14,
      voltage: 14,
      air: 9,
      freeze: 0,
      chaos: 0,
    },
    {
      playStyle: 1,
      difficulty: 1,
      level: 7,
      notes: 143,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 31,
      voltage: 29,
      air: 47,
      freeze: 0,
      chaos: 3,
    },
    {
      playStyle: 1,
      difficulty: 2,
      level: 9,
      notes: 188,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 41,
      voltage: 39,
      air: 27,
      freeze: 0,
      chaos: 13,
    },
    {
      playStyle: 1,
      difficulty: 3,
      level: 12,
      notes: 212,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 46,
      voltage: 39,
      air: 54,
      freeze: 0,
      chaos: 19,
    },
    {
      playStyle: 2,
      difficulty: 1,
      level: 7,
      notes: 130,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 28,
      voltage: 29,
      air: 61,
      freeze: 0,
      chaos: 1,
    },
    {
      playStyle: 2,
      difficulty: 2,
      level: 9,
      notes: 181,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 40,
      voltage: 39,
      air: 30,
      freeze: 0,
      chaos: 11,
    },
    {
      playStyle: 2,
      difficulty: 3,
      level: 11,
      notes: 220,
      freezeArrow: 0,
      shockArrow: 0,
      stream: 48,
      voltage: 54,
      air: 27,
      freeze: 0,
      chaos: 41,
    },
  ]

  chartIndex: number = 0

  get isValidSongId() {
    return /^[01689bdiloqDIOPQ]{32}$/.test(this.id)
  }

  async loadSongInfo() {
    if (!this.isValidSongId) return
    try {
      const songInfo = await this.$http.$get<SongInfo>(`/songs/${this.id}`)

      this.name = songInfo.name
      this.nameKana = songInfo.nameKana
      this.nameIndex = songInfo.nameIndex
      this.artist = songInfo.artist
      this.series = songInfo.series
      this.minBPM = songInfo.minBPM
      this.maxBPM = songInfo.maxBPM
    } catch {}
  }
}
</script>
