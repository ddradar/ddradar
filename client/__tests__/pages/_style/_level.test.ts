import type { Api } from '@ddradar/core'
import type { Context } from '@nuxt/types'
import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
} from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'

import { searchCharts } from '~/api/song'
import ChartLevelPage from '~/pages/_style/_level.vue'

jest.mock('~/api/song')

const localVue = createLocalVue()
localVue.use(Buefy)

describe('/_style/_level.vue', () => {
  const $fetchState = { pending: false }
  const stubs = { NuxtLink: RouterLinkStub, ChartList: true }

  describe.each(['single', 'double'])('snapshot test (/%s/19)', style => {
    test('renders correctly', () => {
      // Arrange
      const mocks = { $route: { params: { level: '19', style } }, $fetchState }
      const data = () => ({ charts: [] })
      const wrapper = mount(ChartLevelPage, { localVue, mocks, stubs, data })

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('validate()', () => {
    test.each([
      ['', '1'],
      ['DOUBLE', '1'],
      ['single', ''],
      ['double', 'foo'],
      ['single', '0'],
      ['double', '21'],
      ['single', '-1'],
      ['single', '1.0'],
    ])('/%s/%s returns false', (style, level) => {
      // Arrange
      const wrapper = shallowMount(ChartLevelPage, {
        localVue,
        mocks: {
          $route: { params: { style, level } },
          $fetchState,
        },
        stubs: { NuxtLink: RouterLinkStub, ChartList: true },
      })
      const ctx = ({ params: { style, level } } as unknown) as Context

      // Act - Assert
      expect(wrapper.vm.$options.validate!(ctx)).toBe(false)
    })
    test.each([
      ['single', '1'],
      ['double', '1'],
      ['single', '9'],
      ['double', '9'],
      ['single', '19'],
      ['double', '19'],
      ['single', '20'],
      ['double', '20'],
    ])('/%s/%s returns true', (style, level) => {
      // Arrange
      const params = { style, level }
      const wrapper = shallowMount(ChartLevelPage, {
        localVue,
        mocks: { $route: { params }, $fetchState },
        stubs: { NuxtLink: RouterLinkStub, ChartList: true },
      })
      const ctx = ({ params } as unknown) as Context

      // Act - Assert
      expect(wrapper.vm.$options.validate!(ctx)).toBe(true)
    })
  })

  describe('fetch()', () => {
    test.each([
      ['single', '1', 1, 1],
      ['single', '9', 1, 9],
      ['double', '19', 2, 19],
      ['double', '20', 2, 20],
    ])(
      '/%s/%s calls searchCharts($http, %i, %i)',
      async (style, level, styleExpected, levelExpected) => {
        // Arrange
        const apiMock = mocked(searchCharts)
        const charts: Api.ChartInfo[] = []
        apiMock.mockResolvedValue(charts)
        const $route = { params: { style, level } }
        const $http = {}
        const mocks = { $route, $fetchState, $http }
        const wrapper = shallowMount(ChartLevelPage, { localVue, mocks, stubs })

        // Act
        // @ts-ignore
        await wrapper.vm.$options.fetch!.call(wrapper.vm)

        // Assert
        expect(apiMock).toBeCalledWith($http, styleExpected, levelExpected)
        expect(wrapper.vm.$data.charts).toBe(charts)
      }
    )
  })

  describe('get title()', () => {
    test.each([
      ['single', '1', 'SINGLE 1'],
      ['single', '9', 'SINGLE 9'],
      ['double', '19', 'DOUBLE 19'],
      ['double', '20', 'DOUBLE 20'],
    ])('/%s/%s route returns %s', (style, level, expected) => {
      // Arrange
      const mocks = { $route: { params: { style, level } }, $fetchState }
      const wrapper = shallowMount(ChartLevelPage, { localVue, mocks, stubs })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.title).toBe(expected)
    })
  })
})
