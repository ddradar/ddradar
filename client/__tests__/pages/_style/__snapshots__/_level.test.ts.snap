// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/_style/_level.vue snapshot test (/double/19) renders correctly 1`] = `
<section class="section">
  <h1 class="title">DOUBLE 19</h1>
  <div class="buttons"><a class="button is-info">
      <!----><span>
      1
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      2
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      3
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      4
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      5
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      6
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      7
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      8
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      9
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      10
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      11
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      12
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      13
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      14
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      15
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      16
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      17
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      18
    </span>
      <!---->
    </a><button type="button" to="/double/19" disabled="disabled" class="button is-info is-outlined">
      <!----><span>
      19
    </span>
      <!---->
    </button></div>
  <chartlist-stub charts=""></chartlist-stub>
</section>
`;

exports[`/_style/_level.vue snapshot test (/single/19) renders correctly 1`] = `
<section class="section">
  <h1 class="title">SINGLE 19</h1>
  <div class="buttons"><a class="button is-info">
      <!----><span>
      1
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      2
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      3
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      4
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      5
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      6
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      7
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      8
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      9
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      10
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      11
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      12
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      13
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      14
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      15
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      16
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      17
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      18
    </span>
      <!---->
    </a><button type="button" to="/single/19" disabled="disabled" class="button is-info is-outlined">
      <!----><span>
      19
    </span>
      <!---->
    </button></div>
  <chartlist-stub charts=""></chartlist-stub>
</section>
`;
